statscache/producer.py

```diff
@@ -13,81 +14,94 @@
+                session.rollback()
 
     def make_session(self):
         """ Initiate database connection """
         uri = self.hub.config['statscache.sqlalchemy.uri']
         return statscache.plugins.init_model(uri)
 
+    def buffer(self, message):
+        self.cache.append(message)
+
     def poll(self):
         """
-        Empty the associated bucket and send to each plugin for processing.
+        Empty the cache and distribute the contents to each plugin for
+        processing.
         """
-        now = datetime.datetime.utcnow()
+        timestamp = datetime.datetime.utcnow() # moment cache was cleared
 
-        bucket = self.sister.buckets[self.name]
-        self.sister.buckets[self.name] = []
+        cache = self.cache
+        self.cache = []
 
-        n = len(bucket)
-        log.info("%s called with %i items in the bucket." % (self.name, n))
+        n = len(cache)
+        log.info("%s called with %i cached items." % (type(self).__name__, n))
 
-        for plugin in self.plugins.values():
-            log.info("  Calling %r" % plugin.name)
+        for plugin in self.plugins:
+            log.info("  Calling %r" % plugin.ident)
             session = self.make_session()
             try:
-                plugin.handle(session, bucket)
+                plugin.handle(session, timestamp, cache)
                 session.commit()
             except:
                 log.exception('Error during plugin %r handling.' % plugin)
                 session.rollback()
 
 
-class OneSecondProducer(StatsProducerBase):
-    frequency = statscache.schedule.Schedule(second=range(0, 60, 1))
-
-
-class FiveSecondProducer(StatsProducerBase):
-    frequency = statscache.schedule.Schedule(second=range(0, 60, 5))
-
-
-class OneMinuteProducer(StatsProducerBase):
-    frequency = statscache.schedule.Schedule(second=[0])
-
-
-class OneDayProducer(StatsProducerBase):
-    # Every night at midnight (UTC)
-    frequency = statscache.schedule.Schedule(second=[0], minute=[0], hour=[0])
+def factory():
+    """
+    Producer class factory based on the frequencies of the available plugin
+    classes.
+    """
+    plugins_by_interval = defaultdict(list)
+    for plugin_class in statscache.utils.plugin_classes:
+        interval = plugin_class.interval
+        if isinstance(interval, datetime.timedelta):
+            plugins_by_interval[interval].append(plugin_class)
+
+    for interval, plugin_classes in plugins_by_interval.items():
+        class StatsProducerAnon(StatsProducerBase):
+            """ Dynamically generated class for a specific frequency """
+            pass # we need to programmatically set class attributes
+        frequency = statscache.frequency.Frequency(interval)
+        StatsProducerAnon.frequency = frequency
+        StatsProducerAnon.plugin_classes = plugin_classes
+        StatsProducerAnon.__name__ = 'StatsProducer' + str(frequency)
+        yield StatsProducerAnon
+    return
+
+producers = factory()
```

Comment:
Yes, it is. But, and this is something I should've stated more clearly, it requires [this change](https://github.com/mokshaproject/moksha/pull/29), which is currently only available direct from the git repo. When @ralphbean returns from his vacation, I can work with him to get the version on PyPI updated (he's the package owner).